### ISSUE TITLE
 Milvus server is hung or reponse delay when make large amount of search request with spark 

### ISSUE THREAD

**Issue by beKoala** [milvus-io/milvus#5480]
**Describe the bug**
 The http interface system/status reponse deply or milvus-em cann't be connected or milvus-client cann't be connected

**Steps/Code to reproduce behavior**
Dataset:http://nlp.stanford.edu/data/glove.6B.zip(400,000，50Dim)
Spark Version: 2.4.0
executor number is 40, each executor have 4 cores，the dataset has 50 partitions，each partition has a threadpool with 3 thread s minimum and 5 threads maximum，the batch size of each query request is 10,000

milvus_server lsof over 10,000 the bug must happen

**Expected behavior**

**Method of installation**
- [X] Docker/cpu
- [ ] Docker/gpu
- [ ] Build from source

**Environment details**
- Hardware/Software conditions (OS, CPU, GPU, Memory)
CPU 16 cores
Memory 32GB

- Milvus version (master or released version)
0.10.3-cpu-d091720-f962e8

**Configuration file**
Settings you made in `server_config.yaml` or `milvus.yaml`
version: 0.5

cluster:
  enable: false
  role: rw

general:
  timezone: UTC+8
  meta_uri: mysql://root:milvusroot@172.17.0.6:3306/milvus

network:
  bind.address: 0.0.0.0
  bind.port: 19530
  http.enable: true
  http.port: 19121

storage:
  path: /var/lib/milvus
  auto_flush_interval: 1

wal:
  enable: true
  recovery_error_ignore: true
  buffer_size: 256MB
  path: /var/lib/milvus/wal

cache:
  cache_size: 16GB
  insert_buffer_size: 1GB
  preload_collection:

gpu:
  enable: false
  cache_size: 1GB
  gpu_search_threshold: 1000
  search_devices:
    - gpu0
  build_index_devices:
    - gpu0

logs:
  level: debug
  trace.enable: true
  path: /var/lib/milvus/logs
  max_log_file_size: 1024MB
  log_rotate_num: 0

metric:
  enable: true
  address: 172.17.0.5
  port: 9091

```yaml
paste-file-content-here
```

**Screenshots**

**Additional context**

**Comment by yhmo** [comment]
The description is not very clear, please correct me if I misunderstand:
- you create a collection, dimension=50
- the collection has 50 partitions
- you insert 400,000 vectors into the collection
- client multi-threads to send search request to milvus server, 3 ~ 5 threads send request for one partition, totally there are 150 ~ 150 client threads
- each search request nq=10000

**Comment by beKoala** [comment]
> The description is not very clear, please correct me if I misunderstand:
> 
> * you create a collection, dimension=50
> * the collection has 50 partitions
> * you insert 400,000 vectors into the collection
> * client multi-threads to send search request to milvus server, 3 ~ 5 threads send request for one partition, totally there are 150 ~ 150 client threads
> * each search request nq=10000

-  the collection has only one partition，_default
- executor number is 40, each executor have 4 cores，so totally there are 160 * 3 ~ 160 * 5 client threads

**Comment by yhmo** [comment]
> > The description is not very clear, please correct me if I misunderstand:
> > 
> > * you create a collection, dimension=50
> > * the collection has 50 partitions
> > * you insert 400,000 vectors into the collection
> > * client multi-threads to send search request to milvus server, 3 ~ 5 threads send request for one partition, totally there are 150 ~ 150 client threads
> > * each search request nq=10000
> 
> * the collection has only one partition，_default
> * executor number is 40, each executor have 4 cores，so totally there are 160 * 3 ~ 160 * 5 client threads

After inserted the 400,000 vectors, no more insert? No delete action?

**Comment by beKoala** [comment]
> > > The description is not very clear, please correct me if I misunderstand:
> > > 
> > > * you create a collection, dimension=50
> > > * the collection has 50 partitions
> > > * you insert 400,000 vectors into the collection
> > > * client multi-threads to send search request to milvus server, 3 ~ 5 threads send request for one partition, totally there are 150 ~ 150 client threads
> > > * each search request nq=10000
> > 
> > 
> > 
> > * the collection has only one partition，_default
> > * executor number is 40, each executor have 4 cores，so totally there are 160 * 3 ~ 160 * 5 client threads
> 
> After inserted the 400,000 vectors, no more insert? No delete action?

yes, do nothing.

**Comment by yhmo** [comment]
@beKoala 
I did some test but could not reproduce the hang issue:
1. start milvus 0.10.3 on a 12 cpu cores machine
2. insert 400000 vectors with 50 dimension
3. run client on another machine, open 800 threads to send search request continually, nq=10000

The nq value is pretty large, so each single search cost a couple of seconds. Milvus has a request queue to hold requests, the search engine execute search request one by one, so lot of requests are waiting in the queue, that is why the response returned slowly. At the beginning, response delay will become larger and larger, after a period of time, all response delay will become a constant value, for my test, the value is 337~338 seconds. But milvus didn't hang, the search engine is still keep working.

I check the issues fixed in recent versions, if your code invoke load_collection() and and search() with multi-threads, it could be the issue #4378 which was fixed in v0.10.5.

**Comment by beKoala** [comment]
**@yhmo** 
Although milvus didn't hang, but the response delay.  I turn down the nq to 1000, but the response also delayed. When I stop the Spark task，the server goes well.

Unlike your reproduce, thanks to MilvusGrpcClient can not serialization in java sdk，so I hava to build a new client in every container and every task, such as call() in MapPartitionsFunction. There are lots of clients are connected simultaneous. I look up GrpcServer.cpp find that grpc_concurrency is be set fourfold cpu cores between 32 and 256, does it matter ? How can I use it in my Spark task with only one client ?

**Comment by yhmo** [comment]
> **@yhmo**
> Although milvus didn't hang, but the response delay. I turn down the nq to 1000, but the response also delayed. When I stop the Spark task，the server goes well.
> 
> Unlike your reproduce, thanks to MilvusGrpcClient can not serialization in java sdk，so I hava to build a new client in every container and every task, such as call() in MapPartitionsFunction. There are lots of clients are connected simultaneous. I look up GrpcServer.cpp find that grpc_concurrency is be set fourfold cpu cores between 32 and 256, does it matter ? How can I use it in my Spark task with only one client ?

The grpc_concurrency is be set fourfold cpu cores between 32 and 256, this is to improve throughput, but grpc concurrency also occupy cpu resource, could affect the performance of search request. But the concurrency value is not configurable. That means there are hundreds of channels to send search request to milvus server simultaneously. To reduce the channels count, I think you can build a single client to receive the search data of each Spark task, only this client can send search request to milvus server. Something like:
```
spark_task_1--|
spark_task_2--|
spark_task_3--|====> milvus clien t===> milvus server
......               --|
spark_task_n--|
```

**Comment by yhmo** [comment]
If milvus server is hung, may be similar to #5398. We made a fix in new version v1.1.1 for #5398, this version is released today.

**Comment by shengjun1985** [comment]
It is fixed. 
And if you have more questions, please continue to contact our community.
